README.md
# sharedb-milestone-mongo

[![NPM Version](https://img.shields.io/npm/v/sharedb-milestone-mongo.svg)](https://npmjs.org/package/sharedb-milestone-mongo)
[![Build Status](https://travis-ci.org/share/sharedb-milestone-mongo.svg?branch=master)](https://travis-ci.org/share/sharedb-milestone-mongo.svg?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/share/sharedb-milestone-mongo/badge.svg?branch=master)](https://coveralls.io/github/share/sharedb-milestone-mongo?branch=master)

MongoDB milestone snapshot database adapter for [`sharedb`][1]. Milestone snapshots can be used to speed up the results
of ShareDB's `connection.fetchSnapshot` method by providing points in time on top of which a smaller number of ops can
be applied to reach the requested version.

Milestone snapshots will be stored in a collection called `m_COLLECTION` where `COLLECTION` is the name of your ShareDB
collection.

## Quick start

```javascript
const MongoMilestoneDB = require('sharedb-milestone-mongo');
const ShareDB = require('sharedb');

const milestoneDb = new MongoMilestoneDB('mongodb://localhost:27017/test');
const shareDb = new ShareDB({ milestoneDb: milestoneDb });
```

## Configuration

### Mongo

The underlying Mongo database can be configured in a number of ways. This library uses the [`mongodb`][2]
library, so any configuration that can be used there can be used in this library.

Mongo can be configured simply using a connection string and any desired [options][3]:

```javascript
const milestoneDb = new MongoMilestoneDB('mongodb://localhost:27017/test', { loggerLevel: 'debug' });
```

It can also be configured with a callback that provides an instance of a [`MongoClient`][4]:

```javascript
const mongodb = require('mongodb');

const milestoneDb = new MongoMilestoneDB((callback) => {
  mongodb.connect('mongodb://localhost:27017/test', callback);
});
```

The Mongo connection string or function can be used as the first argument of the constructor, or as part of an options
object:

```javascript
const milestoneDb = new MongoMilestoneDB({
  mongo: 'mongodb://localhost:27017/test',
  loggerLevel: 'debug',
});
```

### Milestone snapshot saving

#### Intervals

By default, ShareDB will save a milestone snapshot with a given frequency. This library defaults to an interval of
1,000, saving milestones when the 1,000th, 2,000th, etc. versions are committed. That interval can be configured:

```javascript
const milestoneDb = new MongoMilestoneDB({
  mongo: 'mongodb://localhost:27017/test',
  interval: 500,
});
```

#### Complex saving logic

If you need more complex saving logic (eg different intervals depending on collection), this can be achieved using
ShareDB middleware:

```javascript
const milestoneDb = new MongoMilestoneDB('mongodb://localhost:27017/test');
const shareDb = new ShareDB({ milestoneDb: milestoneDb });

shareDb.use('commit', (request, callback) => {
  switch (request.collection) {
    case 'foo':
      // Save every 100 versions for collection 'foo'
      request.saveMilestoneSnapshot = request.snapshot.v % 100 === 0;
      break;
    case 'bar':
    case 'baz':
      // Save every 500 versions for collections 'bar' and 'baz'
      request.saveMilestoneSnapshot = request.snapshot.v % 500 === 0;
      break;
    default:
      // Don't save any milestones for collections not named here.
      // IMPORTANT: We have to set this to false to actively disable milestones
      // If left to null, then the default interval will still apply
      request.saveMilestoneSnapshot = false;
  }

  callback();
});
```

Note that the default value of `request.saveMilestoneSnapshot` is `null`. If left to this value, it will use the default
interval logic. If you want to actively disable snapshots, you must make sure to set it to `false`.

### Indexing

By default, indexing is enabled on the milestone collections in order to speed up fetching. This can have an adverse
impact on performance if being enabled on an existing collection that is missing the index. The indexing can be
disabled:

```javascript
const milestoneDb = new MongoMilestoneDB({
  mongo: 'mongodb://localhost:27017/test',
  disableIndexCreation: true,
});
```

## Error codes

### 4100 - Bad request - DB

* 4101 - Must provide valid collection name
* 4102 - Must provide valid ID
* 4103 - Must provide valid snapshot
* 4104 - Must provide valid integer version or null
* 4105 - Must provide valid integer timestamp or null

### 5100 - Internal error - DB

* 5101 - Mongo closed


[1]: https://github.com/share/sharedb
[2]: https://mongodb.github.io/node-mongodb-native/
[3]: https://mongodb.github.io/node-mongodb-native/3.6/api/MongoClient.html#.connect
[4]: https://mongodb.github.io/node-mongodb-native/3.6/api/MongoClient.html


CHANGELOG.md
# 0.5.0

- Depend on ShareDB >= v1.0.0

# 0.4.0

- Add support for [`mongodb@3`](https://mongodb.github.io/node-mongodb-native/3.6/api/)

# 0.3.0

- Skipped version (oops)

# 0.2.0

- Add `fetchSnapshotByTimestamp` supporting methods

# 0.1.0

Initial release


test/factories/factory.js
class Factory {
  static base() { }

  static build(overrideFunction) {
    const instance = this.base();

    if (typeof overrideFunction === 'function') {
      overrideFunction(instance);
    }

    return instance;
  }
}

module.exports = Factory;


test/mongo-milestone-db.js
/* eslint-disable global-require */
const expect = require('chai').expect;
const MongoMilestoneDB = require('../lib/mongo-milestone-db');
const SnapshotFactory = require('./factories/snapshot-factory');

const MONGO_URL = process.env.TEST_MONGO_URL || 'mongodb://localhost:27017/test';

['mongodb2', 'mongodb3'].forEach((driver) => {
  const mongodb = require(driver);

  function create(options, callback) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }

    let db;

    options = Object.assign({
      mongo: (shareDbCallback) => {
        let mongo;

        mongodb.connect(MONGO_URL)
          .then((mongoConnection) => {
            mongo = mongoConnection;
            return MongoMilestoneDB._isLegacyMongoClient(mongo)
              ? mongo.dropDatabase()
              : mongo.db().dropDatabase();
          })
          .then(() => {
            shareDbCallback(null, mongo);
            callback(null, db, mongo);
          });
      },
    }, options);

    db = new MongoMilestoneDB(options);
  }

  describe(`[${ driver } driver]`, () => {
    require('sharedb/test/milestone-db')({create: create});

    describe('MongoMilestoneDB', () => {
      describe('with a mongo callback', () => {
        let db;
        let mongo;

        beforeEach((done) => {
          create((error, createdDb, createdMongo) => {
            if (error) return done(error);
            db = createdDb;
            mongo = MongoMilestoneDB._isLegacyMongoClient(createdMongo)
              ? createdMongo
              : createdMongo.db();
            done();
          });
        });

        afterEach((done) => {
          if (db._mongoPromise) return db.close(done);
          done();
        });

        it('overwrites an existing milestone snapshot', (done) => {
          const id = 'abc';
          const collection = 'testcollection';
          const version = 1;

          const snapshot1 = SnapshotFactory.build((snapshot) => {
            snapshot.id = id;
            snapshot.v = version;
            snapshot.data = {foo: 'bar'};
          });

          const snapshot2 = SnapshotFactory.build((snapshot) => {
            snapshot.id = id;
            snapshot.v = version;
            snapshot.data = {foo: 'baz'};
          });

          db.saveMilestoneSnapshot(collection, snapshot1, (saveError1) => {
            if (saveError1) return done(saveError1);
            db.getMilestoneSnapshot(collection, id, 1, (getError1, retrievedSnapshot1) => {
              if (getError1) return done(getError1);
              expect(retrievedSnapshot1).to.eql(snapshot1);
              db.saveMilestoneSnapshot(collection, snapshot2, (saveError2) => {
                if (saveError2) return done(saveError2);
                db.getMilestoneSnapshot(collection, id, 1, (getError2, retrievedSnapshot2) => {
                  if (getError2) return done(getError2);
                  expect(retrievedSnapshot2).to.eql(snapshot2);
                  done();
                });
              });
            });
          });
        });

        it('adds an index for the snapshots', (done) => {
          const snapshot = SnapshotFactory.build();

          db.saveMilestoneSnapshot('testcollection', snapshot, (saveError) => {
            if (saveError) return done(saveError);
            mongo.collection('m_testcollection').indexInformation((indexError, indexes) => {
              if (indexError) return done(indexError);
              expect(indexes.d_1_v_1).to.be.ok;
              expect(indexes['m.mtime_1']).to.be.ok;
              done();
            });
          });
        });

        it('errors when trying to access a database that is closed', (done) => {
          db.close((closeError) => {
            if (closeError) return done(closeError);
            db.getMilestoneSnapshot('testcollection', 'abc', null, (getError) => {
              expect(getError).to.be.ok;
              done();
            });
          });
        });

        it('defaults to an interval of 1000', () => {
          expect(db.interval).to.equal(1000);
        });
      });

      describe('indexing disabled', () => {
        let db;
        let mongo;

        beforeEach((done) => {
          const options = {
            disableIndexCreation: true,
          };

          create(options, (error, createdDb, createdMongo) => {
            if (error) return done(error);
            db = createdDb;
            mongo = MongoMilestoneDB._isLegacyMongoClient(createdMongo)
              ? createdMongo
              : createdMongo.db();
            done();
          });
        });

        afterEach((done) => {
          db.close(done);
        });

        it('does not add an index for milestones', (done) => {
          const snapshot = SnapshotFactory.build();

          db.saveMilestoneSnapshot('testcollection', snapshot, (saveError) => {
            if (saveError) return done(saveError);
            mongo.collection('m_testcollection').indexInformation((indexError, indexes) => {
              if (indexError) return done(indexError);
              expect(indexes.id_1_v_1).not.to.be.ok;
              done();
            });
          });
        });
      });

      describe('connecting using just URL', () => {
        let db;

        beforeEach(() => {
          db = new MongoMilestoneDB(MONGO_URL);
          return db._db().then(mongo => mongo.dropDatabase());
        });

        afterEach((done) => {
          db.close(done);
        });

        it('can save and fetch milestones', (done) => {
          const collection = 'testcollection';
          const snapshot = SnapshotFactory.build();
          const id = snapshot.id;

          db.saveMilestoneSnapshot(collection, snapshot, (saveError) => {
            if (saveError) return done(saveError);
            db.getMilestoneSnapshot(collection, id, null, (getError, retrievedSnapshot) => {
              if (getError) return done(getError);
              expect(retrievedSnapshot).to.eql(snapshot);
              done();
            });
          });
        });
      });

      describe('a mocked out database that throws all the time', () => {
        let db;

        beforeEach(() => {
          const mockMongo = {
            close: () => {
              throw new Error('Mock: could not close');
            },
            collection: () => {
              throw new Error('Mock: could not get collection');
            },
          };

          db = new MongoMilestoneDB({
            mongo: (callback) => {
              callback(null, mockMongo);
            },
          });
        });

        it('returns a callback with error on close error', (done) => {
          db.close((error) => {
            expect(error).to.be.ok;
            done();
          });
        });

        it('emits an error when saving with no callback', (done) => {
          const snapshot = SnapshotFactory.build();
          db.on('error', (error) => {
            expect(error).to.be.ok;
            done();
          });

          db.saveMilestoneSnapshot('testcollection', snapshot);
        });
      });

      describe('a bad config', () => {
        it('throws when accessing', (done) => {
          const options = {
            mongo: (callback) => {
              callback(new Error('Mock: could not connect'));
            },
          };

          const db = new MongoMilestoneDB(options);
          db.getMilestoneSnapshot('testcollection', 'abc', null, (error) => {
            expect(error).to.be.ok;
            done();
          });
        });
      });

      describe('overriding the interval', () => {
        let db;

        beforeEach(() => {
          const options = {interval: 100};
          db = new MongoMilestoneDB(MONGO_URL, options);
          return db._mongoPromise.then(mongo => mongo.db().dropDatabase());
        });

        afterEach((done) => {
          db.close(done);
        });

        it('has its interval overridden', () => {
          expect(db.interval).to.equal(100);
        });
      });
    });
  });
});


test/factories/snapshot-factory.js
const Factory = require('./factory');

class SnapshotFactory extends Factory {
  static base() {
    return {
      id: 'abc',
      v: 0,
      type: 'json0',
      data: null,
      m: null,
    };
  }
}

module.exports = SnapshotFactory;


lib/mongo-milestone-db.js
const MilestoneDB = require('sharedb').MilestoneDB;
const mongodb = require('mongodb');

class MongoMilestoneDB extends MilestoneDB {
  constructor(mongo, options) {
    if (typeof mongo === 'object') {
      options = mongo;
      mongo = options.mongo;
    }

    // Shallow clone because we delete options later on
    options = MongoMilestoneDB._shallowClone(options) || {};
    super(options);

    this.interval = this.interval || 1000;

    this._disableIndexCreation = options.disableIndexCreation;
    this._milestoneIndexes = new Set();

    // Since we pass our options object straight through to Mongo, we need to remove any non-Mongo
    // options, because these throw warnings, or can even break the connection if setting the
    // validateOptions flag.
    // See: https://github.com/mongodb/node-mongodb-native/blob/bd4fb531a7f599bb6cf50ebbab3b986f191a7ef8/lib/mongo_client.js#L53-L57
    delete options.mongo;
    delete options.interval;
    delete options.disableIndexCreation;
    this._mongoPromise = MongoMilestoneDB._connect(mongo, options);
  }

  close(callback) {
    if (!callback) callback = () => { };

    this._close()
      .then(() => process.nextTick(callback))
      .catch(error => process.nextTick(callback, error));
  }

  saveMilestoneSnapshot(collectionName, snapshot, callback) {
    if (!callback) {
      callback = (error) => {
        if (error) {
          this.emit('error', error);
        } else {
          this.emit('save', collectionName, snapshot);
        }
      };
    }

    if (!collectionName) return process.nextTick(callback, new InvalidCollectionNameError());
    if (!snapshot) return process.nextTick(callback, new InvalidSnapshotError());

    this._saveMilestoneSnapshot(collectionName, snapshot)
      .then(() => {
        process.nextTick(callback, null);
      })
      .catch(error => process.nextTick(callback, error));
  }

  getMilestoneSnapshot(collectionName, id, version, callback) {
    this._getMilestoneSnapshotByVersion(collectionName, id, version)
      .then(snapshot => process.nextTick(callback, null, snapshot))
      .catch(error => process.nextTick(callback, error));
  }

  getMilestoneSnapshotAtOrBeforeTime(collection, id, timestamp, callback) {
    const isAfterTimestamp = false;
    this._getMilestoneSnapshotByTimestamp(collection, id, timestamp, isAfterTimestamp)
      .then(snapshot => process.nextTick(callback, null, snapshot))
      .catch(error => process.nextTick(callback, error));
  }

  getMilestoneSnapshotAtOrAfterTime(collection, id, timestamp, callback) {
    const isAfterTimestamp = true;
    this._getMilestoneSnapshotByTimestamp(collection, id, timestamp, isAfterTimestamp)
      .then(snapshot => process.nextTick(callback, null, snapshot))
      .catch(error => process.nextTick(callback, error));
  }

  _saveMilestoneSnapshot(collectionName, snapshot) {
    return this._collection(collectionName)
      .then((collection) => {
        const query = {d: snapshot.id, v: snapshot.v};
        const updatedSnapshot = MongoMilestoneDB._snapshotToDbRepresentation(snapshot);
        const options = {upsert: true};

        return collection.updateOne(query, {$set: updatedSnapshot}, options);
      });
  }

  _getMilestoneSnapshotByVersion(collectionName, id, version) {
    if (!id) return Promise.reject(new InvalidIdError());
    if (!this._isValidVersion(version)) {
      return Promise.reject(new InvalidVersionError());
    }

    const query = {d: id};
    if (version != null) {
      query.v = {$lte: version};
    }

    const options = {
      sort: {v: -1},
    };

    return this._getMilestoneSnapshotByQuery(collectionName, query, options);
  }

  _getMilestoneSnapshotByTimestamp(collectionName, id, timestamp, isAfterTimestamp) {
    if (!id) return Promise.reject(new InvalidIdError());
    if (!this._isValidTimestamp(timestamp)) {
      return Promise.reject(new InvalidTimestampError());
    }

    const limit = 1;
    let query;
    let sortOrder;

    if (timestamp === null) {
      query = {d: id};
      sortOrder = isAfterTimestamp ? -1 : 1;
    } else {
      const comparator = isAfterTimestamp ? {$gte: timestamp} : {$lte: timestamp};
      query = {
        d: id,
        'm.mtime': comparator,
      };
      sortOrder = isAfterTimestamp ? 1 : -1;
    }

    const options = {
      limit: limit,
      sort: {'m.mtime': sortOrder},
    };

    return this._getMilestoneSnapshotByQuery(collectionName, query, options);
  }

  _getMilestoneSnapshotByQuery(collectionName, query, options) {
    if (!collectionName) return Promise.reject(new InvalidCollectionNameError());

    return this._collection(collectionName)
      .then(collection => collection.findOne(query, options))
      .then(MongoMilestoneDB._databaseRepresentationToSnapshot);
  }

  _close() {
    return this._client()
      .then((client) => {
        this._mongoPromise = null;
        return client.close();
      });
  }

  _client() {
    if (!this._mongoPromise) {
      return Promise.reject(new MongoClosedError());
    }

    return this._mongoPromise;
  }

  _db() {
    return this._client().then(client => (MongoMilestoneDB._isLegacyMongoClient(client)
      ? client
      : client.db()));
  }

  _collection(collectionName) {
    let name;
    let collection;

    return this._db()
      .then((db) => {
        name = MongoMilestoneDB._milestoneCollectionName(collectionName);
        collection = db.collection(name);

        if (this._shouldCreateIndex(name)) {
          // WARNING: Creating indexes automatically like this is quite dangerous in
          // production if we are starting with a lot of data and no indexes
          // already. If new indexes were added or definition of these indexes were
          // changed, users upgrading this module could unsuspectingly lock up their
          // databases. If indexes are created as the first ops are added to a
          // collection this won't be a problem, but this is a dangerous mechanism.
          // Perhaps we should only warn instead of creating the indexes, especially
          // when there is a lot of data in the collection.
          return Promise.all([
            collection.createIndex({d: 1, v: 1}, {background: true, unique: true}),
            collection.createIndex({'m.mtime': 1}, {background: true}),
          ]);
        }

        return Promise.resolve();
      })
      .then(() => {
        if (!this._disableIndexCreation) {
          this._milestoneIndexes.add(name);
        }

        return collection;
      });
  }

  _shouldCreateIndex(milestoneCollectionName) {
    // Given the potential problems with creating indexes on the fly, it might
    // be preferable to disable automatic creation
    return !this._disableIndexCreation
      && !this._milestoneIndexes.has(milestoneCollectionName);
  }

  static _milestoneCollectionName(collectionName) {
    return `m_${ collectionName }`;
  }

  static _connect(mongo, options) {
    if (typeof mongo === 'function') {
      return new Promise((resolve, reject) => {
        mongo((error, db) => {
          error ? reject(error) : resolve(db);
        });
      });
    }

    return mongodb.connect(mongo, options);
  }

  static _snapshotToDbRepresentation(snapshot) {
    const databaseRepresentation = MongoMilestoneDB._shallowClone(snapshot);
    databaseRepresentation.d = databaseRepresentation.id;
    delete databaseRepresentation.id;

    return databaseRepresentation;
  }

  static _databaseRepresentationToSnapshot(databaseRepresentation) {
    if (!databaseRepresentation) {
      return undefined;
    }

    const snapshot = MongoMilestoneDB._shallowClone(databaseRepresentation);
    snapshot.m = snapshot.m == null ? null : snapshot.m;
    snapshot.id = snapshot.d;

    delete snapshot._id;
    delete snapshot.d;

    return snapshot;
  }

  static _shallowClone(object) {
    if (typeof object !== 'object') return object;
    return Object.assign({}, object);
  }

  static _isLegacyMongoClient(client) {
    // mongodb 2.x connect returns a DB object that also implements the
    // functionality of a client, such as `close()`. mongodb 3.x connect
    // returns a Client without the `collection()` method
    return typeof client.collection === 'function'
      && typeof client.close === 'function';
  }
}

class InvalidCollectionNameError extends Error {
  constructor() {
    super();
    this.code = 4101;
    this.message = 'Must provide valid collection name';
  }
}

class InvalidIdError extends Error {
  constructor() {
    super();
    this.code = 4102;
    this.message = 'Must provide valid ID';
  }
}

class InvalidSnapshotError extends Error {
  constructor() {
    super();
    this.code = 4103;
    this.message = 'Must provide valid snapshot';
  }
}

class InvalidVersionError extends Error {
  constructor() {
    super();
    this.code = 4104;
    this.message = 'Must provide valid integer version or null';
  }
}

class InvalidTimestampError extends Error {
  constructor() {
    super();
    this.code = 4105;
    this.message = 'Must provide valid integer timestamp or null';
  }
}

class MongoClosedError extends Error {
  constructor() {
    super();
    this.code = 5101;
    this.message = 'Already closed';
  }
}

module.exports = MongoMilestoneDB;
